### DISCUSSION TITLE
Intel 82599ES not detected

### DISCUSSION THREAD

**Discussion by Phill93** [ipxe/ipxe#553] [Q&A]
Hello,

I have received a few servers with network cards (Supermicro AOC-STGN-i2S) with Intel 82599ES chip (PCI ID 8086:10fb). The chip is supported according to the driver list but ipxe does not find it. ifstat shows only the two 1Gbit onboard (i210).

Can someone help me?

Greetings

Daniel

**Comment by NiKiZe** [comment]
Indeed it is in: https://github.com/ipxe/ipxe/blob/master/src/drivers/net/intelx.c#L471-L485

Have you verified which version of iPXE you are running?

If it is the latest please try building with `DEBUG=intelx` to try and figure out what might cause it to not be detected.

**Comment by Phill93** [comment]
I have complied a version with this options `make bin-x86_64-efi/ipxe.efi DEBUG=intelx` and this is the result

https://user-images.githubusercontent.com/1454797/149347754-e4e5aa6a-e8e6-457b-84ac-dca6de74e07f.mp4

**Comment by Phill93** [comment]
I have now noticed that the problem only occurs in UEFI mode. An iPXE with PCI debugging enabled also only shows the two i210 cards in the debug messages.

**Comment by mcb30** [comment]
This kind of problem is usually caused by broken OEM drivers that fail to disconnect properly when iPXE tries to take over the device.  You can build with `DEBUG=efi_driver` to see the relevant error messages.

We have a [veto mechanism](https://github.com/ipxe/ipxe/blob/master/src/interface/efi/efi_veto.c) in place that can then be used to forcibly unload misbehaving drivers.  The debug output from `DEBUG=efi_driver` should be sufficient to allow a patch for `efi_veto.c` to be created.

**Comment by karcaw** [comment]
Here is a quick patch I needed on my supermicro boxes to get them booting properly:
```
diff --git a/src/interface/efi/efi_veto.c b/src/interface/efi/efi_veto.c
index a3b60d65..d685aa18 100644
--- a/src/interface/efi/efi_veto.c
+++ b/src/interface/efi/efi_veto.c
@@ -494,12 +494,37 @@ efi_veto_vmware_uefipxebc ( EFI_DRIVER_BINDING_PROTOCOL *binding __unused,
        return 1;
 }
 
+static int
+efi_veto_intel_82599 ( EFI_DRIVER_BINDING_PROTOCOL *binding __unused,
+                           EFI_LOADED_IMAGE_PROTOCOL *loaded __unused,
+                           EFI_COMPONENT_NAME_PROTOCOL *wtf __unused,
+                           const char *manufacturer, const CHAR16 *name ) {
+       static const CHAR16 intel82599[] = L"Intel(R) 10GbE Driver 6.7.04 x64";
+       static const char *intel = "Supermicro";
+
+       /* Check manufacturer and driver name */
+       if ( ! manufacturer )
+               return 0;
+       if ( ! name )
+               return 0;
+       if ( strcmp ( manufacturer, intel ) != 0 )
+               return 0;
+       if ( memcmp ( name, intel82599, sizeof ( intel82599 ) ) != 0 )
+               return 0;
+
+       return 1;
+}
+
 /** Driver vetoes */
 static struct efi_veto_candidate efi_vetoes[] = {
        {
                .name = "Ip4Config",
                .veto = efi_veto_ip4config,
        },
+       {
+               .name = "Intel 82599-sfp",
+               .veto = efi_veto_intel_82599,
+       },
        {
                .name = "HP Xhci",
                .veto = efi_veto_hp_xhci,
```

**Reply by NiKiZe** [reply]
Reading https://github.com/ipxe/ipxe/issues/1123#issuecomment-1895418976
Maybe this can be moved over to an PR but fixing the `intel = "Supermicro"` part 😁
